File: api/src/main/java/api/services/UserService.java
package api.services;

import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;


import api.repositories.UserRepository;
import api.dtos.UserDTO;
import api.models.UserModel;
import api.maps.UserMapper;
import api.exceptions.UserModelException;



@Service
public class UserService{
  //tres coupler pour l'instant je suis entrain de checker comment spring boot fait des repo 
  private final UserRepository USER_REPO;//https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html
  private final UserMapper MAPPER;
  //constructor:

  public UserService(UserRepository repo) {
    USER_REPO = repo;
    MAPPER = new UserMapper();
  }
  
  public void Register(UserDTO dto, String psw) throws UserModelException {
     try { 
      BCryptPasswordEncoder encoder = new BCryptPasswordEncoder(16);//to taylor 16 migh be too intense
      String password = encoder.encode(psw);

      USER_REPO.CreateUser(dto.getName(),dto.getEmail(), password);
    
    } catch(Exception ex) {//mieux gerer exception pour repo
      throw new UserModelException("Couldn't find user");
    }//a creer une classe mot de passe pour encapsuler la logic ou sinon quelque fonction
  }




  public void UpdateAccount(UserDTO dto) throws UserModelException{
    try {
      USER_REPO.UpdateUser(dto.getName(),dto.getEmail());

    } catch(Exception ex) {
      throw new UserModelException("Couldn't find user");
    }
  }

  public void DeleteAccount(Long id) throws UserModelException{
    try{
      USER_REPO.DeleteUser(id);
    } catch(Exception ex){
      throw new UserModelException("Couldn't find user");
    }
  }

  public UserDTO GetAccountInfo(Long id) throws UserModelException{ 
    try { 
      System.out.println("REQUEST ID: "+id);
      
      USER_REPO.findAll().forEach(u -> 
      System.out.println("Found user: ID=" + u.getId() + ", Email=" + u.getEmail()));

      System.out.println("after for each");
      UserModel user = USER_REPO.findById(2L)
                                .orElseThrow(()-> new UserModelException("someting"));
 
      return MAPPER.apply(user);
    } catch(Exception ex) {
      throw new UserModelException("Couldn't find user: "+ex.getMessage()); 
    }
  }
}

File: api/src/main/java/api/services/UserService.kt
package api.services

import api.dtos.OptionalUserDto
import api.dtos.UserDTO
import api.dtos.UserLoginDTO
import api.dtos.UserRegisterationDto
import api.entities.UserEntity
import api.exceptions.InvalidCredentialsException
import api.exceptions.UserAlreadyExistsException
import api.exceptions.UserNotFoundException
import api.maps.toDto
import api.maps.toEntity
import api.repositories.UserRepository
import org.springframework.stereotype.Service


@Service
class UserService(
    val userRepo: UserRepository
) {

    fun register(
        registerationData: UserRegisterationDto
    ): UserDTO = if (userRepo.existsByEmail(registerationData.email))
        // User already exists in database so we throw an exception since we cannot register it
        throw UserAlreadyExistsException("User ${registerationData.email} already exists")
    else userRepo.run {
        // FIXME: register should hash the password before saving it to database
        //  Once fixed, remove the line above
        save(
            registerationData.toEntity()
        ).toDto()
    }

    fun login(
        loginData: UserLoginDTO
    ): UserDTO = userRepo.let {
        val registeredUser = it.findByEmailIgnoreCase(loginData.email)
        // FIXME: login should hash the password before comparing it to the one in database
        //  Once fixed, remove the line above
        if (registeredUser.password == loginData.password) registeredUser.toDto()
        else throw InvalidCredentialsException()
    }

    fun findByEmail(email: String): UserDTO = userRepo.findByEmailIgnoreCase(email)
        .toDto()

    fun findByUsername(username: String): List<UserDTO> = userRepo.run {
        findAllByUsernameIgnoreCase(username)
            .map(UserEntity::toDto)
    }

    fun deleteAccount(id: Long) = userRepo.deleteById(id)

    fun update(userId: Long, userUpdateData: OptionalUserDto): UserDTO = userRepo.run {

        val user = findById(userId)
            .orElseThrow { UserNotFoundException("User with id $userId not found") }

        userUpdateData.name?.let { user.username = it }
        userUpdateData.email?.let { user.email = it }
        userUpdateData.password?.let { user.password = it }

        save(user)
            .toDto()
    }

    fun getById(id: Long): UserDTO = userRepo.run {
        findById(id)
            .orElseThrow { UserNotFoundException("User with id $id not found") }
            .toDto()
    }
}